mtc.go

```diff
@@ -2039,3 +2044,68 @@ func (tai *TrustAnchorIdentifier) unmarshal(s *cryptobyte.String) error {
 	tai.Issuer = oid
 	return nil
 }
+
+// VerifyOptions includes parameters for verifying a BikeshedCertificate.
+type VerifyOptions struct {
+	// ValidityWindow is a validity window that covers the certificate. It
+	// is the caller's responsibility to verify the validity window was
+	// signed by the CA, e.g., by verifying the SignedValidityWindow that
+	// contains the validity window.
+	ValidityWindow *ValidityWindow
+
+	// CA includes the parameters of the CA that issued the batch
+	// containing the certificate.
+	CA *CAParams
+
+	// CurrentTime is used to to check if the certificate has expired.
+	CurrentTime time.Time
+}
+
+// Verify is used to verify that a BikeshedCertificate is covered by a validity
+// window. It is the caller's responsibility to verify the validity window was
+// signed by the CA. An error indicates that the certificate does not belong to
+// a batch in the validity window or that the certificate is otherwise invalid.
+func Verify(cert *BikeshedCertificate, opts VerifyOptions) error {
+	p, ok := cert.Proof.(*MerkleTreeProof)
+	if !ok {
+		return fmt.Errorf("Expected MerkleTreeProof; got %t", cert.Proof)
+	}
+
+	// Check if the certificate has expired.
+	if opts.CurrentTime.After(cert.Proof.NotAfter()) {
+		return fmt.Errorf("Certificate has expired (%v)", p.NotAfter().UTC())
+	}
+
+	// Check that the batch number is in range of the validity window.
+	maxBatchNumber := int(opts.ValidityWindow.BatchNumber)
+	minBatchNumber := maxBatchNumber - int(opts.CA.ValidityWindowSize)
+	if minBatchNumber < 0 {
+		minBatchNumber = 0
+	}
+	certBatchNumber := int(p.TrustAnchorIdentifier().BatchNumber)
+	if certBatchNumber < minBatchNumber || certBatchNumber > maxBatchNumber {
+		return fmt.Errorf("Certificate batch number (%d) out of range [%d, %d]",
+			certBatchNumber, minBatchNumber, maxBatchNumber)
+	}
+
+	// Check that the certificate issuer matches the CA.
+	if !bytes.Equal(p.TrustAnchorIdentifier().Issuer, opts.CA.Issuer) {
+		return fmt.Errorf("Certificate issuer (%s) does not match CA (%s)",
+			p.TrustAnchorIdentifier().Issuer, opts.CA.Issuer)
+	}
+
+	// Select the tree head and verify the authentication path.
+	rootIndex := int(HashLen * (maxBatchNumber - certBatchNumber))
+	root := opts.ValidityWindow.TreeHeads[rootIndex : rootIndex+HashLen]
```

Comment:
Better add a function on ValidityWindow to get the root of batch index, which also conveniently checks whether it's in range of that ValidityWindow.

Reply:
Done!